shell/common/asar/archive_win.cc

```diff
@@ -0,0 +1,140 @@
+// Contributors: Weiyun Dai (https://github.com/WeiyunD/), Andrew Lay
+// (https://github.com/guohaolay) Use of this source code is governed by the MIT
+// license that can be found in the LICENSE file.
+
+#include "shell/common/asar/archive.h"
+
+#include <algorithm>
+#include <sstream>
+
+#include "base/logging.h"
+#include "base/strings/string_util.h"
+#include "base/strings/string_util_win.h"
+#include "shell/common/asar/asar_util.h"
+
+using namespace std;
+
+#define INTEGRITY_CHECK_RESOURCE_TYPE L"Integrity"
+#define INTEGRITY_CHECK_RESOURCE_ITEM L"ElectronAsar"
+
+namespace asar {
+
+absl::optional<base::FilePath> Archive::RelativePath() const {
+  // In case user has Windows long path enabled, the exe path should still no
+  // larger than the asar path.
+  size_t buf_len = max(path_.value().size(), (size_t)MAX_PATH);
+  wchar_t* buf = new (std::nothrow) wchar_t[buf_len];
+  if (!buf) {
+    LOG(WARNING) << " " << __func__
+                 << "(): Failed to alloc enough space to store exe file path";
+    return absl::nullopt;
+  }
+
+  DWORD len = GetModuleFileName(NULL, buf, buf_len);
+  wstring exe_path = wstring(buf, len);
+  delete[] buf;
+
+  if (!len) {
+    LOG(WARNING) << " " << __func__
+                 << "(): GetModuleFileName failed. Last error: "
+                 << GetLastError();
+    return absl::nullopt;
+  }
+
+  base::FilePath exe_dir = base::FilePath(exe_path).DirName();
+  base::FilePath relative_path;
+  if (!exe_dir.AppendRelativePath(path_, &relative_path)) {
+    return absl::nullopt;
+  }
+
+  return relative_path;
+}
+
+string GetIntegrityConfig() {
+  // Load integrity config from exe resource
+  HMODULE module_handle = ::GetModuleHandle(NULL);
+
+  HRSRC resource = ::FindResource(module_handle, INTEGRITY_CHECK_RESOURCE_ITEM,
+                                  INTEGRITY_CHECK_RESOURCE_TYPE);
+  if (!resource) {
+    LOG(WARNING) << " " << __func__
+                 << "(): FindResource failed. Last error: " << GetLastError();
+    return "";
+  }
+
+  HGLOBAL rcData = ::LoadResource(module_handle, resource);
+  if (!rcData) {
+    LOG(WARNING) << " " << __func__
+                 << "(): LoadResource failed. Last error: " << GetLastError();
+    return "";
+  }
+
+  auto* res_data = static_cast<const char*>(::LockResource(rcData));
+  int res_size = SizeofResource(NULL, resource);
+
+  if (!res_data || !res_size) {
+    LOG(WARNING) << " " << __func__
+                 << "(): Failed to integrity config from exe resource.";
+    return "";
+  }
+
+  return string(res_data, res_size);
+}
+
+absl::optional<IntegrityPayload> Archive::HeaderIntegrity() const {
+  absl::optional<base::FilePath> relative_path = RelativePath();
+  // Callers should have already asserted this
+  CHECK(relative_path.has_value());
+
+  // Load integrity config from exe resource
+  string integrity_config = GetIntegrityConfig();
+  if (integrity_config.empty()) {
+    LOG(WARNING) << " " << __func__
```

Comment:
I don't think we need `__func__` in these logs, Chrome doesn't do it and the LOG macro already has line numbers and filenames

Reply:
Updated